elastic/beats
# [winlog/winlogbeat] Gracefully handle event channel not found errors

## What does this PR do?

- Added logic to gracefully handle event channel not found errors. This will only apply to event subscriptions and not reading event files (evtx). If a channel not found error is encountered, either during intial open or during reading, the application will attempt to open a subscription to the event after a short delay.
- Added Channel and IsFile methods to the EventLog interface.
- Added IsChannelNotFound function
- Improved logging through further use of structured logging fields.

## Why is it important?

Filebeat/Winlogbeat must be able to resubscribe to a channel if something occurs to the handle. In some cases, the handle may go invalid for some unknown reason, but in these cases, simply closing the handle and resubscribing corrects the issue.

## Checklist


## How to test this PR locally

This is a mildly difficult one to test, although one of the easier ways is to:

1. Ensure Sysmon is not installed on the system
2. Add `Microsoft-Windows-Sysmon/Operational` as an event log source
3. Start filebeat/winlogbeat, observe "channel not found" errors
4. Install Sysmon. Within a few seconds, filebeat/winlogbeat should be able to subscribe to the channel

What I wasn't able to test is if a valid channel that filebeat/winlogbeat was subscribed to went away, saw the error, reinstalled Sysmon, and watch it re-subscribe (this was the issue in both related cases). 

## Related issues

- Closes #30201 
- Relates elastic/sdh-beats#3039

## Use cases

Needed in cases where the requested channel isn't available yet, or if the channel randomly disappears (which has been reported by customers).

## Logs

filebeat:
```json
{"log.level":"error","@timestamp":"2023-02-21T07:47:38.551-0800","log.logger":"input.winlog","log.origin":{"file.name":"winlog/input.go","file.line":113},"message":"Encountered channel not found error when opening Windows Event Log","service.name":"filebeat","id":"winlog-test","input_source":"winlog-test","eventlog":"winlog-test","channel":"Microsoft-Windows-Sysmon/Operational","error":{"message":"The specified channel could not be found."},"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-21T07:47:43.857-0800","log.logger":"eventlog","log.origin":{"file.name":"eventlog/wineventlog.go","file.line":375},"message":"WinEventLog[winlog-test] using subscription query=<QueryList>\n  <Query Id=\"0\">\n    <Select Path=\"Microsoft-Windows-Sysmon/Operational\">*</Select>\n  </Query>\n</QueryList>","service.name":"filebeat","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-21T07:47:43.857-0800","log.logger":"input.winlog","log.origin":{"file.name":"winlog/input.go","file.line":119},"message":"Windows Event Log opened successfully","service.name":"filebeat","id":"winlog-test","input_source":"winlog-test","eventlog":"winlog-test","channel":"Microsoft-Windows-Sysmon/Operational","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-21T07:47:43.857-0800","log.logger":"eventlog_detail","log.origin":{"file.name":"eventlog/wineventlog.go","file.line":417},"message":"WinEventLog[winlog-test] EventHandles returned 65 handles","service.name":"filebeat","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-21T07:47:43.857-0800","log.logger":"eventlog","log.origin":{"file.name":"eventlog/cache.go","file.line":86},"message":"messageFilesCache[winlog-test] size=1","service.name":"filebeat","ecs.version":"1.6.0"}
```

winlogbeat:
```json
{"log.level":"warn","@timestamp":"2023-02-21T08:16:11.411-0800","log.logger":"winlogbeat","log.origin":{"file.name":"beater/eventlogger.go","file.line":143},"message":"Open() encountered channel not found error. Trying again...","service.name":"winlogbeat","id":"Microsoft-Windows-Sysmon/Operational","error":{"message":"The specified channel could not be found."},"channel":"Microsoft-Windows-Sysmon/Operational","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-21T08:16:16.419-0800","log.logger":"eventlog","log.origin":{"file.name":"eventlog/wineventlog.go","file.line":387},"message":"WinEventLog[Microsoft-Windows-Sysmon/Operational] using subscription query=<QueryList>\n  <Query Id=\"0\">\n    <Select Path=\"Microsoft-Windows-Sysmon/Operational\">*</Select>\n  </Query>\n</QueryList>","service.name":"winlogbeat","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-21T08:16:16.419-0800","log.logger":"winlogbeat","log.origin":{"file.name":"beater/eventlogger.go","file.line":150},"message":"Opened successfully.","service.name":"winlogbeat","id":"Microsoft-Windows-Sysmon/Operational","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-21T08:16:16.423-0800","log.logger":"eventlog_detail","log.origin":{"file.name":"eventlog/wineventlog.go","file.line":429},"message":"WinEventLog[Microsoft-Windows-Sysmon/Operational] EventHandles returned 100 handles","service.name":"winlogbeat","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-21T08:16:16.423-0800","log.logger":"eventlog","log.origin":{"file.name":"eventlog/cache.go","file.line":86},"message":"messageFilesCache[Microsoft-Windows-Sysmon/Operational] size=1","service.name":"winlogbeat","ecs.version":"1.6.0"}
```
